### ISSUE TITLE
Computing the pred_features.pickle file

### ISSUE THREAD

**Issue by johnskop7** [EoinKenny/AAAI-2021#2]
Hello,
I would like to ask you what kind of information regarding the data is stored in the pred_features.pickle file and how can i compute this file for a different dataset ? Thank you very much !

**Comment by lyread** [comment]
hi johnskop7，Do you have any idea about pickle File? I also have same Problem.

**Comment by EoinKenny** [comment]
I think I know the answer, and I'm so sorry for the late reply. I'm also sorry for not having code to reproduce that, I have no idea what I was thinking.

The pred_features.pickle file is basically the saved data of all the activations in the second last layer of the CNN (before softmax).

Say the layer is 128 dimensions, and we have 10 classes, we want to save the training data activations in this layer for each class, so you basically have a tensor of shape (10, n, 128), where n is the number of training instances predicted as that class. We iterate the whole training data and collect it.

It looks like in the code I save this as a nested dictionary, with ['class']['activations'] opening up a 2d tensor here:

```
def get_data_for_feature(dist_data, target_class, feature_map_num):
	data = np.array(dist_data[target_class]['activations'])
	data = data.T[feature_map_num].T.reshape(data.shape[0],1)
	return data
```

So, to save this file, you would need to do the following:

```
pred_features = dict()
for i in range(training_data):
        pred, activations = cnn(training_data[i])
        pred_features[pred][activations].append(activations)

```
Then you should get it. It's not exactly what I did I'm sure, I probably also saved the label in the dict as [pred][label] or something like that, but this should be enough to run the code now.

Basically, you want to get a tensor into the HurdleModel class as `data` which is a 1d tensor of all the training data activations for one neuron for one predicted class.

Hope that makes sense?
